docs/docs/optimizing-performance.md

```diff
@@ -0,0 +1,228 @@
+
+## Tricky Examples
+
+Let's say that you have a component that just renders the string passed in `props.value`. We could implement this with `React.PureComponent`:
+
+```javascript
+class StringComponent extends React.PureComponent {
+  render() {
+    return <div>{this.props.value}</div>;
+  }
+}
+```
+
+If your components use more complex data structures, this might not be so simple. For example, let's say your component renders a comma-separated list of words. This code does *not* work correctly:
+
+```javascript
+class ListOfWords extends React.PureComponent {
+  render() {
+    return <div>{this.props.words.join(',')}</div>;
+  }
+}
+```
+
+The problem is that `PureComponent` will do a shallow comparison between the old and new values of `this.props.words`. If your application has a sequence of events like:
+
+1. `let words = ['alpha', 'beta', 'gamma'];`
+2. Render `<ListOfWords words={words} />`
+3. `words.push('delta')`
+4. Re-render `<ListOfWords words={words} />`
+
+The `PureComponent` logic will actually prevent the rerender during step 4. This happens because the old and new values of `words` are the same list. `PureComponent` just does a shallow comparison, so it mistakenly thinks that nothing has changed in step 4.
+
+One way to solve this is by writing a more complicated `shouldComponentUpdate` function:
+
+```javascript
+class ListOfWords extends React.Component {
+  render() {
+    return <div>{this.props.words.join(',')}</div>;
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    // We have to do a possibly-expensive deep equality check
+    if (props.words.length !== nextProps.words.length) {
+      return true;
+    }
+    for (let i = 0; i < props.words.length; i++) {
+      if (props.words[i] !== nextProps.words[i]) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+```
+
+Since the `ListOfWords` component relies on data deep within the props, we have to to a deep comparison to implement `shouldComponentUpdate`. This may not improve performance at all. The fundamental problem is that we want to quickly check for equality of large data structures, without doing an expensive iteration to compare them.
+
+## The Power Of Not Mutating Data
+
+If you never mutate data structures, you can use shallow comparisons and get the same result as a deep comparison. In this case, the code `words.push('delta')` is mutating the `words` array. Instead of using `push`, you can update without mutating by using `concat` or the ES6 spread operator:
+
+```js
+// The mutable way. This breaks React.PureComponent
+words.push('delta');
+
+// Update words without mutating. This works with React.PureComponent
+words = words.concat(['delta']);
+
+// Update words without mutating, with slightly nicer syntax
+words = [...words, 'delta'];
+```
```

Comment:
Let's show how this looks together with `setState` from a parent component.
Otherwise people will write `this.state.words = [...this.state.words, 'delta']`.